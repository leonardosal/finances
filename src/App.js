import React from 'react';

import Home from './containers/Home';
import Header from './components/Header';

import './styles/global.scss';

const App = () => (
  <>
    <Header />
    <Home />
  </>
);

export default App;
