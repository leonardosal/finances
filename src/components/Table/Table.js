import React from 'react';

import propTypes from 'prop-types';

import './table.scss';

import Row from '../Row';

const Table = ({ transactions }) => (
  <>
    {transactions.map(item => (
      <Row
        key={item.id}
        type={item.type}
        description={item.description}
        value={item.value}
      />
    ))}
  </>
);

Table.propTypes = {
  transactions: propTypes.arrayOf(
    propTypes.shape({
      id: propTypes.string,
      type: propTypes.string,
      description: propTypes.string,
      value: propTypes.string,
    })
  ),
};

Table.defaultProps = {
  transactions: [],
};

export default Table;
