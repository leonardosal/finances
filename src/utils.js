import uuidv4 from 'uuid/v4';

export const TYPE = {
  DEBIT: 'DEBIT',
  CREDIT: 'CREDIT',
};

export const formatValue = value => {
  return parseInt(value, 10).toLocaleString('pt-br', {
    minimumFractionDigits: 2,
    style: 'currency',
    currency: 'BRL',
  });
};

export const calculateBalance = transactions => {
  return transactions.reduce((acc, current) => {
    if (current.type === TYPE.DEBIT) {
      return parseInt(acc, 10) - parseInt(current.value, 10);
    }
    return parseInt(acc, 10) + parseInt(current.value, 10);
  }, 0);
};

export const saveInStorage = transactions => {
  window.localStorage.setItem('@transactions', JSON.stringify(transactions));
};

export const loadFromStorage = () => {
  const strTransactions = window.localStorage.getItem('@transactions');
  const transactions = JSON.parse(strTransactions);
  return transactions;
};

export const addItem = (list, item) => {
  return [{ id: uuidv4(), ...item }, ...list];
};
