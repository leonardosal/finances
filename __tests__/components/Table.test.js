import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Table from '../../src/components/Table';
import { TYPE } from '../../src/utils';

it('should render Table component with correct list of transactions', () => {
  const transactions = [
    {
      id: 'A',
      description: 'Uber',
      type: TYPE.DEBIT,
      value: 17.0,
    },
    {
      id: 'B',
      description: 'Salary',
      type: TYPE.CREDIT,
      value: 2000.0,
    },
  ];
  const component = mount(<Table transactions={transactions} />);
  expect(toJson(component)).toMatchSnapshot();
});
