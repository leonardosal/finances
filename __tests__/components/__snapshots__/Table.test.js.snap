// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Table component with correct list of transactions 1`] = `
<Table
  transactions={
    Array [
      Object {
        "description": "Uber",
        "id": "A",
        "type": "DEBIT",
        "value": 17,
      },
      Object {
        "description": "Salary",
        "id": "B",
        "type": "CREDIT",
        "value": 2000,
      },
    ]
  }
>
  <Row
    description="Uber"
    key="A"
    type="DEBIT"
    value={17}
  >
    <div
      className="row"
    >
      <span>
        Uber
      </span>
      <span
        className="debit"
      >
        R$17,00
      </span>
    </div>
  </Row>
  <Row
    description="Salary"
    key="B"
    type="CREDIT"
    value={2000}
  >
    <div
      className="row"
    >
      <span>
        Salary
      </span>
      <span
        className="credit"
      >
        R$2.000,00
      </span>
    </div>
  </Row>
</Table>
`;
